### ISSUE TITLE
Feedback Wanted!

### ISSUE THREAD

**Issue by ekristen** [ekristen/aws-nuke#389]
Please provide your feedback. Here are some of the issues.

- #298 
- #388

**Comment by ekristen** [comment]
@eaudetcobello I'm considering adding a feature to flip from exclusion to inclusion via https://github.com/ekristen/aws-nuke/issues/327 -- there's much to consider with adding this behavior.

**Comment by BwL1289** [comment]
Exclude all AWS managed IAM policies as well as resources created by CDK. Essentially, reset the AWS account to how it was at initial creation.

I can't tell from the documentation whether all AWS managed IAM policies are automatically ignored.

**Comment by ekristen** [comment]
@BwL1289 and @eaudetcobello you can sort of already filter for removal using [Global Filters Feature](https://ekristen.github.io/aws-nuke/features/global-filters/) combined with the `invert` flag, but it would only work with a single tag, with [Filter Groups](https://ekristen.github.io/aws-nuke/config-filtering/#filter-groups) you could possible extend this further.

**Comment by BwL1289** [comment]
@ekristen a technique I'm using for CDK is:

```yaml
presets:
  # Exclude cdk resources from being nuked
  cdk:
    filters:
      # Exclude resources with cdk in the name
      __global__: # Catch all for all resources
        - type: "contains"
          value: "cdk"
        - type: glob
          value: "*cdk*"
        - property: Name
          type: "glob"
          value: "*cdk*"
        - property: ARN
          type: "glob"
          value: "*cdk*"
        - property: tag:aws:cloudformation:stack-name
          value: "CDKToolkit"
```

Should this work as I expect?

**Comment by BwL1289** [comment]
As I'm finalizing CDK template, two other feature requests:

1. Filter default AWSServiceRoles and their associated IAM Policies
2. Filter S3 Bucket Policies

**Comment by ekristen** [comment]
Default service roles are already filtered.

Bucket policy isn't a separate resource and I'm not sure there's an argument to make it one. Why would you want to nuke a bucket policy but not the bucket?

**Comment by BwL1289** [comment]
> Default service roles are already filtered

Thanks. Is this in the docs? Happy to help add it if not.

> Why would you want to nuke a bucket policy but not the bucket?

I wouldn't. Similar to `IAMRoleAttachmentPolicy`, I want to retain the bucket policy along with the Bucket. Looking at `S3Bucket` source code, this may be default behavior as well.

**Comment by ekristen** [comment]
@nimaft you can only run against a single account at a time, but you can configure as many accounts as you want in the configuration file.

**Comment by Xylon2** [comment]
I tried using this but it's very complicated to configure and use. I ran it one time and it seemed to delete it's own access key and didn't even delete my VPCs which was all I needed it for.

**Comment by zack-is-cool** [comment]
It would be nice to have a clear expectation of what properties you can filter on each resource type.

Today I was attempting to exclude a CloudTrailTrail using tags. [CloudTrail objects have tags](https://docs.aws.amazon.com/awscloudtrail/latest/APIReference/API_Tag.html). But I don't believe this tool will allow you to filter on tags? see: https://github.com/ekristen/aws-nuke/blob/main/resources/cloudtrail-trails.go#L60-L63

Please correct me if I'm wrong 😄 

I'll submit an issue if you confirm we should be able to filter on this

thanks in advance @ekristen

**Comment by ekristen** [comment]
@zack-is-cool unfortunately not all resources support both properties and even if does not all have their tags exposed. PRs welcome! I am currently working on auto-documentation based on property data when it exists, upcoming releases will have better documentation, there is a PR in progress https://github.com/ekristen/aws-nuke/pull/391 

Please feel free to open an issue for the cloud trail objects missing tags.

**Comment by jarrodallan** [comment]
I hope this is the right thread, apologies if not.

A feature I would appreciate when deleting resources, specifically CloudFormation, triggering the remove but not waiting for the delete completion. CloudFormation delete should hopefully take care of the majority of the cleanup itself but it takes a decent amount of time to delete every resource in a stack.

Perhaps a method where the delete is triggered for all stacks then each is checked once the deletes have been triggered, rather than waiting for the full deletion before starting to delete other resources.

**Comment by ekristen** [comment]
@jarrodallan async deletes are in the works.

**Comment by benriou** [comment]
Feature request : could you please consider supporting wildcards in resource-types.exclusions ?

This would avoid to list any existing sub-service on a given AWS service.

Before : 
```
resource-types:
  excludes:
    # These services are not sold by AWS anymore
    - "OpsWorksInstance"
    - "OpsWorksCMServer"
    - "OpsWorksCMBackup"
    - "OpsWorksApp"
    - "OpsWorksUserProfile"
    - "OpsWorksLayer"
    - "OpsWorksCMServerState"
```
 
After : 
```
resource-types:
  excludes:
    # These services are not sold by AWS anymore
    - "OpsWorks*"
```

**Comment by ekristen** [comment]
@benriou this is covered via https://github.com/ekristen/aws-nuke/issues/302 and there is now a PR pending merge today to address it https://github.com/ekristen/aws-nuke/pull/479

**Comment by ekristen** [comment]
Any additional feedback please open an issue or comment on an existing one. I'm closing this issue now. Thank you everyone.
